### DISCUSSION TITLE
Changing response based on params/header

### DISCUSSION THREAD

**Discussion by onordgren** [directus/directus#12573] [Archive]
Hey!

Anyone know if it's possible to change the output of a api call based on stuff passed in the request, like params or headers? 

For example, say that I have a collection with a field called body and a translation relation where each translation has a language and a body field. I would then like to be able to call the end point and give a param like locale=fr-FR that would return the collection but with the French version of the body instead of the main one. 

What I have tried is to add two hooks, one with middleware before that reads the query param and adds the appropriate translation fields to the query. Then I have a items read filter that maps the response and picks the body from the translation and replaces the main body with that. 

This is working fine but what I'm lacking is a way to communicate between the middleware and the read so that this only happens when there is a locale=xx-YY in the request params. What I did first was to add a dummy field to the query but after adding field permissions this didn't work anymore.

Is there any cleaner way of doing this?

**Comment by rijkvanzanten** [comment]
The read filter hooks don't currently have access to the request information, as the call to read may or may not be done from a request in the first place. I think you're current approach might be the next best thing (where you dynamically update the used query based on another request flag) 🤔

**Reply by onordgren** [reply]
My code is something like this, the problem I currently have is knowing if to apply the changes in filter `items.read` or not. Previously I added a dummy field called `_locale` to `req.query.fields` in the middleware but after adding permissions to fields this is not really feasible anymore.

```js
const hook = ({ init, filter, action }) => {
  init('middlewares.before', ({ app }) => {
    app.use((req, _res, next) => {
      const { locale } = req.query;
      if (locale) {
        req.query = {
          ...req.query,
          fields: [...(req.query.fields ?? ['*']), 'translations.*'],
          filter: {
            translations: {
              languages_code: locale,
            },
          },
        };
      }
      return next();
    });
  });

  filter('items.read', async (input, meta, context) => {
    const shouldMapTranslations = true // This is where I would like to figure our if to apply the mapping or not
    if (shouldMapTranslations) {
      meta.query.fields;
      return [...input].map((item) => {
        const { translations, ...rest } = item;
        const { id, markets_code, languages_code, ...translation } =
          translations[0];
        return {
          ...rest,
          ...translation,
        };
      });
    }
    return input;
  });
};

module.exports = hook;
```

**Reply by rijkvanzanten** [reply]
You might be able to add a new property to query itself, rather than relying on fields? Slightly hacky maybe, but it'd get the job done

f.e.

```
req.query.__locale = 'en-US';
```
